src/clients/FtpGateway.cc

```diff
@@ -1042,9 +1042,9 @@ Ftp::Gateway::checkAuth(const HttpHeader * req_hdr)
     /* Test URL login syntax. Overrides any headers received. */
     loginParser(request->url.userInfo(), true);
 
-    /* name is missing. that's fatal. */
-    if (!user[0])
-        fatal("FTP login parsing destroyed username info");
+    /* empty user is not an error; if someone is intent on passing that to
+     * us, go ahead and pass it to the server; this will almost certainly
+     * fail but at least fail gracefully */
```

Comment:
This comment contradicts Ftp::Gateway::loginParser() logic (at least [^1]): There, the logic is documented as (I am paraphrasing for clarity): if the username is empty, retain what we have already. Since this code starts with a non-empty "anonymous" value, I believe this fatal() call should be unreachable in correctly written code. The fact that it is reachable, implies (to me), that there is a bug somewhere -- somewhere we are erasing the previously set username.

This is especially bad because this method logic is a sequence of username (and password) extraction steps, with the last successful step winning. The fact that this fatal() code is reachable means there is also a danger that we are erasing a non-empty non-default username somewhere.

Instead of forwarding what looks like garbage and contradicting ourselves, would it be possible to fix Ftp::Gateway::loginParser() so that it decides whether username is present (and updates `user` field) _after_ rfc1738_unescape()? AFAICT, that simple change would address the actual bug here. After that, this if+fatal() statement should be replaced with Assure().

[^1]: The new comment also contradicts the intent behind the buggy fatal() call if that intent was to assert that our code did not overwrite the default "anonymous" value.

Reply:
Hmhm. I think we could keep it consistent with the rest of the code by simply adding a `|| strlen(login.c_str()) == 0` check to the `Ftp::Gateway::loginParser` to prevent empty usernames from overwriting anything that came before.

That being said, I think that this consistency is bought by continueing to not treat the special case of an empty username correctly (see @MegaManSec comment about empty usernames). Looking at the Uri class, I see that ftp://@<host> is eventually treated the same as ftp://<host> with regard to username -- "anonymous" is used, although the former example provides a username "".  `Ftp::Gateway::loginParser` checks the size of the `SBuf &login` and return directly if it's empty. That fails for the offending `AG00` base64 sequence because for that check SBuf is not treated as a c-string and leaves us with the `'\0` username that is not expected.

Thus, by following the intention of the current code and fixing the Parser in that way, we accept that an empty username equals "no username", which could be ok given that "empty username" probably does not have an overwhelming relevance in practice.

Please let me know what you think

PS: A fix to actually handle the empty username correctly seems a bit more of an effort since it would involve the Uri Parser, because the distinction between empty and none is already killed there, AFAICS.

Reply:
> I think that this consistency is bought by continueing to not treat the special case of an empty username correctly

I bet you are right.

> A fix to actually handle the empty username correctly seems a bit more of an effort

Agreed. That correct handling cannot be implemented well without introducing new types and changing quite a bit of code, including low-quality FTP code.

My recommendation for _this_ PR is to preserve consistent (but incorrect) treatment of empty user names (or what we decide, due to lack of appropriate support in code, is an equivalent of an empty user name). When/if somebody decides to improve Squid FTP compliance with regard to empty user names, they can post a PR dedicated to that improvement.

I am resolving this change request and opening another one to discuss your code change proposal.

P.S. Just to avoid a misunderstanding: My primary goal here is to make sure that this PR does not make things worse (e.g., by adding a comment that is not consistent with the rest of the code). I am not suggesting or insisting on any specific solution to address (or work around) the vulnerability.